### PR TITLE
[core] Prevent obfuscating `Pair` type

### LINKED ISSUE #26531: SDK 50 and LinearGradientView does not work with start and end coordinates
### Minimal reproducible example

https://snack.expo.dev/@hnzufx/simple-linear-gradient

### Summary

In Expo SDK 50 you cannot set start and end coordinates. You will get the error:

 ERROR  Cannot set prop 'endPoint' on view 'class expo.modules.lineargradient.LinearGradientView'
→ Caused by: java.lang.IllegalArgumentException: Class declares 0 type parameters, but 2 were provided.
 ERROR  Cannot set prop 'startPoint' on view 'class expo.modules.lineargradient.LinearGradientView'
→ Caused by: java.lang.IllegalArgumentException: Class declares 0 type parameters, but 2 were provided.

You can reproduce the bug by creating a new project:

npx expo install expo-linear-gradient
cd expo-linear-gradient

and the linked snack as App.js. The snack works because it's SDK 49. In SDK 50 it doesn't work.

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 13.5.2
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.11.0 - /usr/local/opt/node@20/bin/node
      Yarn: 1.22.21 - /usr/local/bin/yarn
      npm: 10.2.4 - /usr/local/opt/node@20/bin/npm
      Watchman: 2023.12.04.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
    IDEs:
      Android Studio: 2021.1 AI-211.7628.21.2111.8193401
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.2 => 50.0.2 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.2 => 0.73.2 
    npmGlobalPackages:
      eas-cli: 6.0.0
      expo-cli: 6.3.10
    Expo Workflow: managed

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes #26531
Closes ENG-11137
Proguard is obfuscating the `Pair` type causing the `PairTypeConverter` to lose type information. This leads to this error 
`java.lang.IllegalArgumentException: Class declares 0 type parameters, but 2 were provided.` when the app is running in release. The other converters seem fine. Confirmed by inspecting the dex classes.

# How
~Added a rule to keep the names of types implementing the `Serializable` interface.~
After discussing with @kudo narrowed it to just adding the rule to keep the `Pair` type

# Test Plan
- Expo Go running in release
- NCL 
- Using the provided repro

### CODE REVIEW

File: packages/expo-modules-core/android/proguard-rules.pro [hunk 1]
```diff
@@ -9,6 +9,7 @@
 -keep enum * implements expo.modules.kotlin.types.Enumerable {
   *;
 }
+-keepnames class * implements java.io.Serializable
```
Comment (Kudo) [inline packages/expo-modules-core/android/proguard-rules.pro:12]: having a sense that Serializable is too broad and may keep user's code if they are using Serializable.
- is it possible to keep only expo-module's method parameters?
- or at least maybe we could keep Pair first

Comment (alanjhughes) [reply packages/expo-modules-core/android/proguard-rules.pro:12]: We could just target `Pair` but if we add more types, it'll be the same issue. It's also mentioned [here](https://www.guardsquare.com/manual/configuration/examples) under `Processing serializable classes
`

Comment (alanjhughes) [reply packages/expo-modules-core/android/proguard-rules.pro:12]: Do the rules here not just affect `expo-modules-core` or are they merged with the others in the app?

Comment (Kudo) [reply packages/expo-modules-core/android/proguard-rules.pro:12]: i was thinking that would affect the app, please correct me if i'm wrong.

Comment (alanjhughes) [reply packages/expo-modules-core/android/proguard-rules.pro:12]: I'm not sure to be honest but I'm also not sure if it's a big issue if it does affect the app module. We just don't want to obfuscate the class name. Keeping just `Pair` will work but we'll need to remember to test the rules again if we add more converters.

Comment (Kudo) [reply packages/expo-modules-core/android/proguard-rules.pro:12]: let's investigate whether "keep only expo-module's method parameters" is feasible.

**Review by lukmccall — APPROVED** [review]

**Pull request merged**